torch/nn/utils/_per_sample_grad.py

```diff
@@ -38,20 +43,33 @@ def call_for_per_sample_grads(module, batch_size, loss_reduction="mean"):
         rewrites that wrap an `nn.Linear` module. See Opacus for an example
     """
 
-    def maybe_build_expanded_weight(og_tensor):
+    def maybe_build_expanded_weight(og_tensor, batch_size):
         if og_tensor.requires_grad:
             return ExpandedWeight(og_tensor, batch_size, loss_reduction)
         else:
             return og_tensor
 
+    def compute_batch_size(*args, **kwargs):
+        args_and_kwargs = tree_flatten(args)[0] + tree_flatten(kwargs.values())[0]
```

Comment:
This imposes a new restriction that args and kwargs contains pytree-able data types. For function transforms (vmap) we assume that and that's fine, but in general this will not work with arbitrary nn.Modules.

But tbh this should work in most cases

Reply:
Hmmm fair. If it's not pytree-able, it will still work if the user explicitly gives a batch size because it won't call this function. I'm not seeing an `is_pytreeable` function but if I put this in a try/catch, I should be able to give the user an error message to that effect